nutmeg release notes (#2056)
openedx/edx-documentation: en_us/open_edx_release_notes/source/nutmeg.rst

* the `frontend-learner-portal-base`_ library
* The `xblock.fragment module`_ and deprecated :code:`id_generator` method parameters in :code:`xblock.runtime`
* The legacy courseware experience (rendered server-side by Django) will be removed. The Learner MFE will be required.
* EdxRestApiClient is no longer supported, as you may have been able to tell from the many, many deprecation warnings.
* DraftModuleStore (also know as Old Mongo Modulestore) will be removed. "Old Mongo" course access was already removed in nutmeg.
* microsites djangoapp
* the ability to import legacy OLX attributes :code:`slug`, :code:`name`, :code:`id` (discussion block), :code:`for`, and :code:`attempts`

.. _bokchoy test suites: https://github.com/openedx/public-engineering/issues/13
.. _frontend-learner-portal-base: https://github.com/openedx/frontend-learner-portal-base/issues/31
.. _xblock.fragment module: https://github.com/openedx/public-engineering/issues/15
.. _microsites djangoapp: https://github.com/openedx/public-engineering/issues/69
.. _import legacy OLX attributes: https://github.com/openedx/public-engineering/issues/74

Future deprecations and removals
--------------------------------

> Major deprecation work is being funded between now and the Olive release, scheduled for December 2022. Please review the `DEPR: Deprecation & Removal`_ board on Github to be sure you have stopped using deprecated technologies.

Reviewer: ```suggestion
.. note:: Major deprecation work is being funded between now and the Olive release, scheduled for December 2022. Please review the `DEPR: Deprecation & Removal`_ board on Github to be sure you have stopped using deprecated technologies.
```

Small suggestion to put this in a `note` directive to make it more visible.

> .. _DEPR\: Deprecation & Removal: https://github.com/orgs/openedx/projects/9/views/4

Author: FYI @feanil 

I also tried to pull a list of the DEPR tickets currently in removing, just to emphasize what is going away.

Reviewer: Looks great!

---------
The Dates Tab has been added as a default static tab on all courses. All new courses will automatically include the Dates Tab. In order to properly have the Dates Tab show up for all your existing courses, a backfill course tabs management command has been created.

.. code-block:: shell

    $ python manage.py cms backfill_course_tabs

Weekly Course Goals
-------------------
The old course goals feature has been replaced with a new weekly learning goals feature. Users set a goal for how frequently they want to learn per course and get reminder emails about their goals. See `4.30. Enabling the Weekly Learning Goals Feature in "Installing, Configuring, and Running the Open edX Platform"`_ for instructions on how to configure this feature and more details on how the feature works. The new weekly learning goals feature is controlled with the same flag as the previous course goals feature.

Authentication Micro-Frontend (MFE)
-----------------------------------

The Authentication Micro-Frontend is available as an experimental feature, but support for theming, internationalization and path-based deployments are incomplete. It will be supported in the next release, Olive.

> 

Author: @kdmccormick can I / should I say more about the authn MFE? 

Ideally, I think I'd like to point to a tag or branch, and a setting to turn it on. But I haven't been able to find either of those.

Reviewer 2: Linking to https://github.com/openedx/frontend-app-authn makes sense

However I'm not sure how it's available as an experimental feature if it's not tagged - I guess there is a tutor plugin for it?

Reviewer 3: There is no support for the authn MFE out of the box in Tutor. (I think we should mention that)

LTI Support
-----------

* Course authors can now define static and dynamic custom parameters that will be sent to the LTI Tool Provider at launch. See the `Custom LTI Parameter section of the LTI Consumer XBlock Readme`_ for more details.
* Mobile apps can now support LTI by opening it in a browser

.. _Custom LTI Parameter section of the LTI Consumer XBlock Readme: https://github.com/openedx/xblock-lti-consumer/blob/master/README.rst#custom-lti-parameters

Gradebook Micro-frontend (MFE)
------------------------------

* Added support for transifex translations.
* Added support for custom theming.

==========================
Administrators & Operators
==========================

* Various improvements and bugfixes have been applied to `Tutor`_, the officialy-supported Open edX distribution and installation method. Notable features include an overhauled Tutor Plugin API and a new CLI for mounting repositories during development. You can see the full list by viewing `Tutor's changelog, starting at v13.0.1`_ (the first Tutor release after Maple) and reading upwards until v14.0.0 (the first Tutor release supporting Nutmeg).

> .. _Tutor's changelog, starting at v13.0.1: https://github.com/overhangio/tutor/blob/master/CHANGELOG.md#v1301-2021-12-20

Author: @kdmccormick @regisb is this Ok? Or do you want to suggest more details?

.. _earlier releases: https://edx.readthedocs.io/projects/edx-developer-docs/en/latest/named_releases.html
.. _Open edX Platform: https://open.edx.org

.. contents::
 :depth: 1
 :local:

================
Breaking Changes
================

SafeSessionMiddleware
---------------------
Before upgrade: Check that your logs do not contain warnings starting with "SafeCookieData user at request", or that these warnings are very rare. If they are common, see the SafeSessionMiddleware section below in the Administrators and Operators Section.

Django Admin login disabled by default
--------------------------------------
Default Django admin login window is disabled and now one has to login from LMS.

===================
Learner Experiences
===================

User Tours
----------
User Tours are walkthroughs that can be presented to users in micro-frontends (MFEs). The default tours that exist are: "Course Home New User Tour" (`screencast`_), "Course Home Existing User Tour", and "Courseware New User Tour".

> In order for User Tours to properly work, the backpopulate user tours management command should be run.

Reviewer 3: Should the command be run just once? Or frequently? (I hope just once)

> In order for User Tours to properly work, the backpopulate user tours management command should be run.

Author: @Dillon-Dumesnil or @MatthewPiatetsky can you help answer these questions about the course tours management command?

.. _screencast: https://user-images.githubusercontent.com/25124041/143145608-6886237d-ea83-42a4-ac2c-555b07392723.mov

.. code-block:: shell

    $ python ./manage.py lms backpopulate_user_tours

Dates Tab
---------
The Dates Tab has been added as a default static tab on all courses. All new courses will automatically include the Dates Tab. In order to properly have the Dates Tab show up for all your existing courses, a backfill course tabs management command has been created.

.. code-block:: shell

>     $ python manage.py cms backfill_course_tabs

Author: @Dillon-Dumesnil or @MatthewPiatetsky another management command question: 

> Will this [backfill_course_tabs] management command add the date tab to every old course, or will it only make it possible to add the tab to older courses?

The Dates Tab has also been removed from the legacy learner experience. It is only available in the Learner MFE. This is the first step in remove the legacy learner experience altogether, which will be completed in the Olive release.

Weekly Course Goals
-------------------
The old course goals feature has been replaced with a new weekly learning goals feature. Users set a goal for how frequently they want to learn per course and get reminder emails about their goals. See `4.30. Enabling the Weekly Learning Goals Feature in "Installing, Configuring, and Running the Open edX Platform"`_ for instructions on how to configure this feature and more details on how the feature works. The new weekly learning goals feature is controlled with the same flag as the previous course goals feature.

Authentication Micro-Frontend (MFE)
-----------------------------------

The Authentication Micro-Frontend is available as an experimental feature, but support for theming, internationalization and path-based deployments are incomplete. It will be supported in the next release, Olive.

Discussions
-----------

> The Discussions Micro-Frontend (MFE) is available as an experimental feature, but support for theming, internationalization and path-based deployments are incomplete. It will be supported in the next release, Olive.

Reviewer 2: Similarly here I might do
```
`Discussions Micro-Frontend (MFE)`_

.. _Discussions Micro-Frontend (MFE): https://github.com/openedx/frontend-app-discussions
```

> Learning Micro-Frontend (MFE)

Reviewer 2: I'm surprised to see this as a top-level section when there were other Learning MFE changes above. I think I'd pull all Learning MFE changes into one section, with a few bigger subsections for the larger items and an "other features" section for these smaller bullets.

Author: I'm not sure what sections you're referring to. I guess the Dates Tab is technically part of the Learning MFE, but what else?

It's also not clear to me how to add another level of hierarchy with the rst template we have.

-----------------------------

* When an XBlock fails to render, the user will not see an error message "An unexpected error occurred. Please click the button below to refresh the page."
* RTL languages are now supported
* Support for the experimental Discussions MFE and "Course Live" feature was added

Account Micro-frontend (MFE)
----------------------------

> The Account Micro-frontend is now enabled by default. The legacy account pages will be removed in the next release, Olive.

Reviewer 2: I think a link here would be good - do we have docs on this MFE? or link the repo, if the README is good?

Payment Micro-frontend

> -----------------------

Reviewer 2: Is this section a work in progress?

Author: apparently I have nothing to say about the payment mfe. I've removed it.

Mobile Experience
-----------------

* Added support for notices in apps.

> * Added an activation opt in, opt out checkbox in the registration page.

Reviewer 2: I don't know what this means. Could you elaborate a little? Also what's the default state of the checkbox?

* Mobile apps can now support LTI by opening it in a browser

Special Exams Experience
------------------------

> To take a proctored exam, the learner must now be enrolled in a :code:`verified` course track. Note that ID verification (IDV) is not required.

Reviewer 2: What was the previous behavior?

======================
Instructor Experiences
======================

> * Course tabs have a new, standard order. When course authors change the order of static tabs, the ordering of default tabs will be enforced for a more uniform location across courses.

Reviewer 2: Would this be an accurate rewording?

Default course tabs have a new, standard order. Course authors may still change the order of their custom static tabs, but the ordering of the default tabs cannot be changed. This ensures a more uniform location across courses.

Author: That's my understanding.

* When setting grading policies, course authors can now set the minimum grade cutoff to 99. Previously it could not be higher than 97.

> * When exporting teams csv, username and external key are now separate columns (previously, an external key would override the username, leading to ambiguity when importing teams csv)

Reviewer 2: I am not sure what this means. I think this could be clearer if location is called out, possibly "When exporting the teams csv *from the X section of the instructor dashboard*, username and..."

Author: I've never used this feature, but apparently there is a page for managing teams. I cribbed some details from https://edx.readthedocs.io/projects/edx-partner-course-staff/en/latest/course_features/teams/teams_setup.html#managing-teams

Course Authoring Micro-frontend (MFE)
-------------------------------------

The Course Authoring MFE is available as an experimental feature, but support for theming, internationalization and path-based deployments are incomplete. It will be supported in the next release, Olive.

The Course Authoring MFE replaces the "Pages" menu item in Studio with "Pages and Resources" and gives access to new features, like the "Course Live xBlock" for Zoom integration.

Open-Response Assessments (ORA)
-------------------------------

The ORA Grading Micro-frontend (MFE) is available as an experimental feature, but support for theming, internationalization and path-based deployment maybe be incomplete. It will be support in the next release, Olive. The MFE can be enabled by setting a value for :code:`ORA_GRADING_MICROFRONTEND_URL`.

LTI Support
-----------

> * Can now add custom parameter template values in order to be substituted by a processor. See `PR 215`_ for configuration details.

Reviewer 2: Gotta be honest I've not much idea what this means or if it's useful to me, even after skimming the PR. I think this would be clearer and more helpful in the form of:

"USER_TYPE (instructor?) can now add custom parameter template values"

it would be even more clear and helpful for there to be a _why_ here - why would you want to do this? The PR does not really answer the question for me.

Author: I found some docs on this, which help a little. I rewrote it and linked to the docs.

* Mobile apps can now support LTI by opening it in a browser

.. _PR 215: https://github.com/openedx/xblock-lti-consumer/pull/215

> Gradebook Micro-FrontEnd

Reviewer 2: ```suggestion
Gradebook Micro-frontend (MFE)
```

------------------------

* Added support for transifex translations.
* Added support for custom theming.

==========================
Administrators & Operators
==========================

> * An internal performance improvement called “learning sequences” has been opt-in for a few releases, but is now always-on for Nutmeg. If you have any courses that have not been re-published on Koa or later, run the :code:`simulate_publish` cms django command on your courses before upgrading, to populate the learning sequence data.

Reviewer 2: can "learning sequences" link to documentation?

.. code-block:: shell

    $ python manage.py cms simulate_publish

* By setting the :code:`ENABLE_NEW_RELIC` to false, the new relic logging integration will be disabled across all MFEs.

> * It's now possible to add optional, but not toggled, extra fields to the registration form. See the `PR and example config`_ for more information.

Reviewer 2: Maybe I'm just tired at the end of the day but I'm not grokking these PRs. I don't understand what "toggled" means in this context (what does it mean to add a field, make it optional, and then "toggle" it or not "toggle" it?)

Author: I removed the word "toggle" and tried to describe the use case. I also added a link to the relevant documentation for the setting in question.

.. _PR and example config: https://github.com/openedx/edx-platform/pull/29047

> * Add another field to the registration form, :code:`marketing_emails_opt_in`, for marketing email opt-in.

Reviewer 2: is this field default-on for all instances? can you turn it off? I assume "marketing emails opt-in" is kind of an edx.org thing.

Author: It defaults to checked here as well. I think it can be removed by removing it from the `REGISTRATION_EXTRA_FIELDS` setting, but I'm not certain.

* Bug fix: When using GMSTP (Gmail) for sending bulk email, retriable SMTP exceptions were not caught and bulk sending failed. This has been fixed.

Bulk Course Email Tool
----------------------

* Added the ability to filter recipients of bulk course emails based on the last_login date of Users enrolled in a course run. This feature can be enabled by setting a value for the :code:`BULK_COURSE_EMAIL_LAST_LOGIN_ELIGIBILITY_PERIOD` setting. Its value should be an integer (representing months) that represents the eligibility period from the current date to receive a message. The new setting defaults to None which keeps this new feature disabled (and there will be no change in behavior in how recipients are filtered/selected for a message).

* Added a simple :code:`bulk_email_disabledcourse` table that allows for the bulk email tool to be disabled for specific course runs, even if the bulk email flag is on and the course is enabled in the :code:`bulk_email_courseauthorization` table. A course team will not be able to see the bulk email tab on the instructor dashboard for whatever course runs are in this table.

> * the setting :code:`EMAIL_USE_DEFAULT_FROM_FOR_BULK` was changed to :code:`EMAIL_USE_COURSE_ID_FROM_FOR_BULK`. The behavior was also changed, such that those who wish to use their course id  in the from address for bulk email must now enable the flag to true.These changes were made in order to avoid non existent from address to fail in email servers. (In progress?)

Reviewer 2: ```suggestion
* The setting :code:`EMAIL_USE_DEFAULT_FROM_FOR_BULK` was changed to :code:`EMAIL_USE_COURSE_ID_FROM_FOR_BULK`. Additionally the behavior was also changed. Previously, it was possible for a "from" address constructed out of a course id to be non existent, and thus fail in email servers. Now, those who wish to use their course id  in the "from" address for bulk email must enable the flag to true. 
```

Took me awhile to grok this one but I think this is what you're saying?

SafeSessionMiddleware rejects mismatching requests and sessions
---------------------------------------------------------------

> Background: This is an existing middleware that provides several protections against vulnerabilities that could result from cache misconfigurations or other bugs resulting in one user getting a different user's session.

Reviewer 2: ```suggestion
Background: SafeSessionMiddleware is an existing middleware that provides several protections against vulnerabilities that could result from cache misconfigurations or other bugs resulting in one user getting a different user's session.
```

makes it clearer

Changed: Previously if a user mismatch was detected between request or session and response, the middleware would log warnings; now, it will invalidate the session and send an error response. The toggle :code:`ENFORCE_SAFE_SESSIONS` is enabled by default, but can be disabled to return to just log warnings.

> Before upgrade: Check that your logs do not contain warnings starting with "SafeCookieData user at request", or that these warnings are very rare. If they are common, there is likely a false positive caused by some custom login, masquerading, or registration code that needs to call mark_user_change_as_expected. Otherwise, valid requests may be rejected.

Reviewer 2: There's no tip here about what to do if they have these common ones. Should they resolve the false positives and do the upgrade? Should they just not enforce safe sessions?

Author: I presume the tip is to edit your custom code to call `mark_user_change_as_expected`

Migrations
----------

> (there's no notes in https://openedx.atlassian.net/wiki/spaces/COMM/pages/3205201949/Nutmeg on migrations. I presume that does not mean there are none. Is there a standard way to do migrations that we could write out here?)

Reviewer: Suggestion:

There are no known migrations that will cause compatibility issues when deployed.  As always migrations should be run before the new code is deployed.

* `learner_dashboard.enable_masters_program_tab_view <https://edx.readthedocs.io/projects/edx-platform-technical/en/latest/featuretoggles.html#featuretoggle-learner_dashboard.enable_masters_program_tab_view>`_
* `learner_dashboard.enable_program_tab_view <https://edx.readthedocs.io/projects/edx-platform-technical/en/latest/featuretoggles.html#featuretoggle-learner_dashboard.enable_program_tab_view>`_
* `learner_dashboard.enable_program_tab_view <https://edx.readthedocs.io/projects/edx-platform-technical/en/latest/featuretoggles.html#featuretoggle-learner_dashboard.enable_program_tab_view>`_
* `new_core_editors.use_new_problem_editor <https://edx.readthedocs.io/projects/edx-platform-technical/en/latest/featuretoggles.html#featuretoggle-new_core_editors.use_new_problem_editor>`_
* `new_core_editors.use_new_text_editor <https://edx.readthedocs.io/projects/edx-platform-technical/en/latest/featuretoggles.html#featuretoggle-new_core_editors.use_new_text_editor>`_
* `new_core_editors.use_new_video_editor <https://edx.readthedocs.io/projects/edx-platform-technical/en/latest/featuretoggles.html#featuretoggle-new_core_editors.use_new_video_editor>`_

the following settings were removed:

> * ENABLE_COMPREHENSIVE_THEMING

Author: Oops. That's not a good one to remove by mistake. 

I should probably review my methodology for building this list.

* agreements.enable_integrity_signature

* course_experience.latest_update

* course_goals.number_of_days_goals

* course_home.course_home_use_legacy_frontend

* learner_dashboard.enable_program_discussions

* learning_sequences.use_for_outlines

* request_utils.capture_cookie_sizes

* user_authn.redirect_to_microfrontend

Dependency updates
------------------

> (any significant dependency updates?)

Reviewer 3: Node was indeed upgraded to 16 for the MFEs, but not for edx-platform. I do not think of this as a "dependency update", on the same level as Mysql/Mongodb upgrade because it does not require complex data migrations. But I'm probably wrong about that, as people will have to upgrade their custom MFEs.

.. _frontend-learner-portal-base: https://github.com/openedx/frontend-learner-portal-base/issues/31
.. _xblock.fragment module: https://github.com/openedx/public-engineering/issues/15
.. _microsites djangoapp: https://github.com/openedx/public-engineering/issues/69
.. _import legacy OLX attributes: https://github.com/openedx/public-engineering/issues/74

Future deprecations and removals
--------------------------------

.. note:: Major deprecation work is being funded between now and the Olive release, scheduled for December 2022. Please review the `DEPR: Deprecation & Removal`_ board on Github to be sure you have stopped using deprecated technologies.

.. _DEPR\: Deprecation & Removal: https://github.com/orgs/openedx/projects/9/views/4

=============================
Researcher & Data Experiences
=============================

* added a :code:`complete_video` event that fires when a user has watched a video to the end. Requires the waffle switch :code:`completion.enable_completion_tracking`

=====================
Developer Experiences
=====================

* support for custom xBlock editors in Studio
* added an API for updating user's email opt-in setting

> * It is now possible to reference a plugin context in Instructor Dashboard templates.

Reviewer 2: do you have links for any of these? Particularly the custom editors & the plugin context. Both seem bigger than a brief sentence

Events and Filters Extension Framework
--------------------------------------

> Core extensibility: added a new way of extending the core through `Open edX Events & Filters`_ (part of `OEP-50: Hooks Extension Framework`_)

Reviewer: ```suggestion
Core extensibility: We have added a new way of extending the core through `Open edX Events & Filters`_ (part of `OEP-50: Hooks Extension Framework`_)
```

.. code-block:: shell

    $ python manage.py cms backfill_course_tabs

The Dates Tab has also been removed from the legacy learner experience. It is only available in the Learner MFE. This is the first step in remove the legacy learner experience altogether, which will be completed in the Olive release.

Weekly Course Goals
-------------------
The old course goals feature has been replaced with a new weekly learning goals feature. Users set a goal for how frequently they want to learn per course and get reminder emails about their goals. See `4.30. Enabling the Weekly Learning Goals Feature in "Installing, Configuring, and Running the Open edX Platform"`_ for instructions on how to configure this feature and more details on how the feature works. The new weekly learning goals feature is controlled with the same flag as the previous course goals feature.

Authentication Micro-Frontend (MFE)
-----------------------------------

> The `Authentication Micro-Frontend (MFE)`_ is available as an experimental feature, but is not yet supported in Tutor. It will be supported in the next release, Olive.

Reviewer 2: @pdpinch could we move all the experimental features under one section, and also call them `pre-alpha`? Experimental to me implies there's a flag or setting that will turn them on. Beta even sounds more complete than where we're currently at. Enabling one of these MFEs would take a lot of work and it's undocumented. I recommend this structure:

```
Pre-Alpha Features
-------------------

The following Micro-frontends (MFEs) are in a "pre-alpha" state. They exist on GitHub but are not yet supported in Tutor. Additionally, they may lack key features such as support for theming, internationalization, and path-based deployments. We include mention of them because we expect all of them to be supported in the next release, Olive.

* `Authentication Micro-Frontend (MFE)`_
* `Discussions Micro-Frontend (MFE)`_
* `Course Authoring Micro-frontend (MFE)`_

.. _Authentication Micro-Frontend: https://github.com/openedx/frontend-app-authn
.. _Discussions Micro-Frontend (MFE): https://github.com/openedx/frontend-app-discussions
.. Course Authoring Micro-frontend (MFE): TBD
```

and please put this sub-section at the very bottom of this section (ie under Special Exams).

Author: I agree with this approach, but I don't think they all belong within "Learner Experiences." 

Given the challenge of enabling them, I put this at the end of the "Administrators & Operators" section.

.. _Authentication Micro-Frontend: https://github.com/openedx/frontend-app-authn

Discussions
-----------

The `Discussions Micro-Frontend (MFE)`_ is available as an experimental feature, but support for theming, internationalization and path-based deployments are incomplete. It will be supported in the next release, Olive.

.. _Discussions Micro-Frontend (MFE): https://github.com/openedx/frontend-app-discussions

Learning Micro-Frontend (MFE)
-----------------------------

* When an XBlock fails to render, the user will now see an error message "An unexpected error occurred. Please click the button below to refresh the page."
* RTL languages are now supported

> * Support for the experimental Discussions MFE was added

Reviewer 2: ```suggestion
* Support for the pre-alpha Discussions MFE was added
```

======================
Instructor Experiences
======================

* Course tabs have a new, standard order. When course authors change the order of static tabs, the ordering of default tabs will be enforced for a more uniform location across courses.
* When setting grading policies, course authors can now set the minimum grade cutoff to 99. Previously it could not be higher than 97.
* When exporting teams csv, username and external key are now separate columns (previously, an external key would override the username, leading to ambiguity when importing teams csv)

Course Authoring Micro-frontend (MFE)
-------------------------------------

The Course Authoring MFE is available as an experimental feature, but support for theming, internationalization and path-based deployments are incomplete. It will be supported in the next release, Olive.

The Course Authoring MFE replaces the "Pages" menu item in Studio with "Pages and Resources" and gives access to new features, like the "Course Live xBlock" for Zoom integration.

Open-Response Assessments (ORA)
-------------------------------

> The ORA Grading Micro-frontend (MFE) is available as an experimental feature, but support for theming, internationalization and path-based deployment maybe be incomplete. It will be support in the next release, Olive. The MFE can be enabled by setting a value for :code:`ORA_GRADING_MICROFRONTEND_URL`.

Reviewer 2: It seems like ORA can remain as "experimental" because you _can_ enable it with a setting. Unless it's not supported by Tutor... if that's the case, it should be in "beta" or "pre-alpha" (not sure which) with a mention that Tutor does not support it.

Author: I don't think it's enough to change the setting. I moved it to the pre-alpha section.
